src/odemis/dataio/catmaid.py

```diff
@@ -0,0 +1,360 @@
+
+    if project_id is None or stack_id is None:
+        try:
+            token, _, _ = read_config_file(base_url, token=True)
+            token = "5306d14006f8b71841766b9385e8bd3712820d04"
+            auth = CatmaidApiTokenAuth(token) if token else None
+            project_url = "{url}/projects/".format(url=base_url)
+            response = requests.get(project_url, auth=auth)
+            if response.status_code == 401:
+                raise AuthenticationError("Wrong token while getting project info at {}".format(project_url))
+            else:
+                response.raise_for_status()
+            project_info = response.json()
+            if project_info == [] and token is None:
+                raise AuthenticationError(
+                    "No project at {}, this Catmaid instance does not contain projects"
+                    "or a token should have been provided.".format(project_url))
+            project_id = project_info[0]["id"] if project_id is None else project_id
+            # loop through the projects to get the info of the project matching the project id.
+            project = [p for p in project_info if p["id"] == project_id]
+            stack_id = project[0]["stacks"][0]["id"] if stack_id is None else stack_id
+        except Exception:
+            # Try with project id and stack id is 1 if the project info cannot be accessed.
+            project_id = 1
+            stack_id = 1
+        logging.info(
+            "Project id and/or stack id not entered, using project {} and stack {}.".format(project_id, stack_id))
+    return AcquisitionDataCatmaid(base_url, project_id, stack_id)
+
+
+class DataArrayShadowPyramidalCatmaid(DataArrayShadow):
+    """
+    This class implements the read of a Pyramidal Catmaid instance.
+    """
+
+    def __init__(self, stack_info, base_url):
+        """
+        Constructor
+        stack_info (dict): information about the Catmaid stack and tiles in the stack.
+        base_url (str): URL where the Catmaid instance is hosted.
+        dtype (numpy.dtype): the data type
+        metadata (dict str->val): The metadata
+        """
+        shape = (stack_info["dimension"]["x"], stack_info["dimension"]["y"])
+        tile_shape = (stack_info["mirrors"][0]["tile_width"], stack_info["mirrors"][0]["tile_height"])
+        # resolution is given in nanometers by catmaid.
+        resolution = {k: v * 1e-9 for k, v in stack_info["resolution"].items()}
+        metadata = dict()
+        metadata[model.MD_PIXEL_SIZE] = (resolution["x"], resolution["y"])
+        dtype = numpy.uint8
+
+        if stack_info["num_zoom_levels"] == -1:
+            # if num_zoom_levels is -1 the maxzoom must be estimated by using: 2**maxzoom = max_dim / min_size
+            # Rewriting this results in maxzoom = log(max_dim / min_size) / log(2)
+            max_dim = max(shape)
+            min_size = min(tile_shape)
+            maxzoom = int(math.ceil(math.log(max_dim / min_size) / math.log(2)))
```

Comment:
In Python, log() has an extra argument for the base, so: `log(x)/log(2)` -> `log(x, 2)`